src/librustc_mir/borrow_check/nll/type_check/mod.rs

```diff
@@ -684,6 +684,25 @@ impl<'a, 'b, 'gcx, 'tcx> TypeVerifier<'a, 'b, 'gcx, 'tcx> {
                         }
                     }
                 }
+                ty::Generator(def_id, substs, _) => {
```

Comment:
Why is this needed? Do you need to downcast to access upvars now?

Reply:
Using `state_tys` during NLL seems problematic too.

Reply:
Hmm, I don't think this was actually needed, I found it during grep and added it for "completeness". That was probably wrong in retrospect.

Reply:
The previous MIR type check ran after generator transformation, so maybe this does too? It would be needed in that case at least.

Reply:
I think it should run before, so you can remove this and check. cc @nikomatsakis

Reply:
It runs before generator transformation.

Reply:
Okay, I can remove it and replace it with a comment then.